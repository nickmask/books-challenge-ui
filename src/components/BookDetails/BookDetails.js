import React from "react";
import PropTypes from "prop-types";

import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import { fetchBookDetails } from "../../utils/api";

import styles from "./BookDetails.module.css";

class BookDetails extends React.Component {
  state = {
    book: {
      author: "",
      description: "",
      id: "",
      price: "",
      stockAmount: "",
      thumbnail: "",
      title: ""
    },
    isFetching: true,
    error: null
  };

  componentDidMount = async () => {
    if (!this.state.book.author) {
      try {
        const bookId = this.props.match.params.id;
        const book = await fetchBookDetails(bookId);

        this.setState({ isFetching: false, book });
      } catch (error) {
        alert("Error getting book details. Please try again later");
        this.setState({
          error
        });
      }
    }
  };

  onAddToCart = () => {
    this.props.onAddToCart(this.state.book);
  };

  isStockRemaining = () => {
    const { id, stockAmount } = this.state.book;

    const totalStock = stockAmount;
    const totalInCart = this.props.cartItems.filter(item => id === item.id)
      .length;
    const remainingStock = totalStock - totalInCart;

    return remainingStock > 0;
  };

  render() {
    const {
      author,
      description,
      price,
      stockAmount,
      thumbnail,
      title
    } = this.state.book;
    const { isFetching } = this.state;

    return (
      <div className={styles.container}>
        {isFetching ? (
          <LoadingSpinner />
        ) : (
          <div>
            <h1 className={styles.pageHeader} aria-label="Book title">
              {title}
            </h1>
            <hr className={styles.separator} />
            <img
              className={styles.bookCover}
              src={thumbnail}
              alt={`Cover of ${title} book`}
            />
            <div>
              <p aria-label="Book author">{`${author.firstName} ${
                author.lastName
              }`}</p>
              <p aria-label="Book description">{description}</p>
              <p aria-label="Book price">Â£{price}</p>
              <p aria-label="Stock amount">
                {stockAmount > 0
                  ? `Available stock: ${stockAmount}`
                  : "Out of stock"}
              </p>
              <button
                type="submit"
                disabled={!this.isStockRemaining()}
                onClick={this.onAddToCart}
              >
                Add to cart
              </button>
              <button
                className={styles.buyNow}
                disabled={stockAmount <= 0}
                type="submit"
                onClick={() => alert("To be implemented")}
              >
                Buy now
              </button>
            </div>
          </div>
        )}
      </div>
    );
  }
}

BookDetails.propTypes = {
  onAddToCart: PropTypes.func.isRequired,
  cartItems: PropTypes.array.isRequired
};

export default BookDetails;
