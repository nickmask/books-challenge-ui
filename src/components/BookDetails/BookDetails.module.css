.container {
  padding: 0.5rem 1rem;
  margin: 0.5rem 0 0 0;
  background: #fff;
  box-shadow: 0.1rem 0.1rem 0.2rem rgba(0, 0, 0, 0.1);
}

.bookCover {
  object-fit: cover;
  width: 200px;
  height: 305px;
}

.buyNow {
  margin-left: 20px;
}

.separator {
  display: block;
  height: 1px;
  border: 0;
  border-top: 1px solid #f0f0f0;
  margin: 0.5em 0;
  padding: 0;
}

.pageHeader {
  margin-bottom: 0;
}
