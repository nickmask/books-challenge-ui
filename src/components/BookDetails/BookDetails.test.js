import React from "react";
import { shallow } from "enzyme";
import createRouterContext from "react-router-test-context";

import BookDetails from "./BookDetails";

describe("BookDetails", () => {
  const context = createRouterContext();
  const book = {
    id: "c1dc42ab-548f-4f96-8350-2f56303d7532",
    title: "A Life Changed",
    author: {
      id: "b6545db7-2a50-4532-8225-3fec60470502",
      firstName: "John",
      lastName: "Doe"
    },
    price: 10,
    description:
      "Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do.",
    stockAmount: 1,
    thumbnail:
      "https://images.unsplash.com/photo-1496484805162-42c00f79cdce?dpr=1&auto=format&fit=crop&w=568&h=379&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D",
    isbn: "0000044280374"
  };

  const cartItems = [
    { id: "c1dc42ab-548f-4f96-8350-2f56303d7532" },
    { id: "c1dc42ab-548f-4f96-8350-2f56303d7532" }
  ];

  const onAddToCart = jest.fn();

  const match = {
    params: {
      id: "c1dc42ab-548f-4f96-8350-2f56303d7532"
    }
  };

  const props = { cartItems, onAddToCart, match };

  it("should render correctly", async () => {
    const component = shallow(<BookDetails {...props} />, {
      context,
      disableLifecycleMethods: true
    });

    component.setState({
      isFetching: false,
      book
    });

    expect(component).toMatchSnapshot();
  });

  it("isStockRemaining should return false when no stock is remaining", () => {
    const component = shallow(<BookDetails {...props} />, {
      context,
      disableLifecycleMethods: true
    });

    component.setProps({ cartItems });

    component.setState({
      isFetching: false,
      book: {
        id: "c1dc42ab-548f-4f96-8350-2f56303d7532",
        title: "A Life Changed",
        author: {
          id: "b6545db7-2a50-4532-8225-3fec60470502",
          firstName: "John",
          lastName: "Doe"
        },
        price: 10,
        description:
          "Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do.",
        stockAmount: 2,
        thumbnail:
          "https://images.unsplash.com/photo-1496484805162-42c00f79cdce?dpr=1&auto=format&fit=crop&w=568&h=379&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D",
        isbn: "0000044280374"
      }
    });

    const isStockRemaining = component.instance().isStockRemaining;

    expect(isStockRemaining()).toEqual(false);
  });

  it("isStockRemaining should return true when stock is remaining", () => {
    const component = shallow(<BookDetails {...props} />, {
      context,
      disableLifecycleMethods: true
    });

    component.setProps({ cartItems });

    component.setState({
      isFetching: false,
      book: { ...book, stockAmount: 4 }
    });

    const isStockRemaining = component.instance().isStockRemaining;

    expect(isStockRemaining()).toEqual(true);
  });

  it("should disable 'add to cart' when no stock is remaining", () => {
    const component = shallow(<BookDetails {...props} />, {
      context,
      disableLifecycleMethods: true
    });

    component.setProps({ cartItems });

    component.setState({
      isFetching: false,
      book: { ...book, stockAmount: 2 }
    });

    const addToCartButton = component.find('button[children="Add to cart"]');

    expect(addToCartButton.prop("disabled")).toEqual(true);
  });

  it("should 'add to cart' is not disabled when stock is remaining", () => {
    const component = shallow(<BookDetails {...props} />, {
      context,
      disableLifecycleMethods: true
    });

    component.setProps({ cartItems });

    component.setState({
      isFetching: false,
      book: { ...book, stockAmount: 5 }
    });

    const addToCartButton = component.find('button[children="Add to cart"]');

    expect(addToCartButton.prop("disabled")).toEqual(false);
  });
});
