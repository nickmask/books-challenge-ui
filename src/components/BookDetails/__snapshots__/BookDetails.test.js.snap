// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails should render correctly 1`] = `
<div
  className="container"
>
  <div>
    <h1
      aria-label="Book title"
      className="pageHeader"
    >
      A Life Changed
    </h1>
    <hr
      className="separator"
    />
    <img
      alt="Cover of A Life Changed book"
      className="bookCover"
      src="https://images.unsplash.com/photo-1496484805162-42c00f79cdce?dpr=1&auto=format&fit=crop&w=568&h=379&q=60&cs=tinysrgb&ixid=dW5zcGxhc2guY29tOzs7Ozs%3D"
    />
    <div>
      <p
        aria-label="Book author"
      >
        John Doe
      </p>
      <p
        aria-label="Book description"
      >
        Nulla sunt sunt pariatur officia et esse fugiat proident est. Ad laboris ullamco adipisicing nisi in ex. Magna voluptate culpa velit exercitation nisi ipsum est do.
      </p>
      <p
        aria-label="Book price"
      >
        Â£
        10
      </p>
      <p
        aria-label="Stock amount"
      >
        Available stock: 1
      </p>
      <button
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Add to cart
      </button>
      <button
        className="buyNow"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Buy now
      </button>
    </div>
  </div>
</div>
`;
